### PR TITLE
Test noise-handshake with converted keys

### PR DESCRIPTION
**Pull request by pavel-mukhanov**
Introduce method to convert our consensus keys(Ed25519) to keys suitable for Noise-handshake with `XK` pattern(Curve25519). 
```
crypto::convert_keypair_for_handshake(..)
```
Conversion is based on `libsodium` methods:
```
crypto_sign_ed25519_pk_to_curve25519
crypto_sign_ed25519_sk_to_curve25519
```

### CODE REVIEW

File: exonum/src/events/noise/tests.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        let (public_key_i, secret_key_i) = gen_keypair();
+        let (public_key_r, secret_key_r) = gen_keypair();
+
+        // Convert to Curve25519 keys.
+        let (public_key_i, secret_key_i) =
+            convert_keypair_for_handshake(public_key_i, secret_key_i).unwrap();
+        let (public_key_r, secret_key_r) =
+            convert_keypair_for_handshake(public_key_r, secret_key_r).unwrap();
+
+        // Do DH
```
Comment (popzxc) [inline exonum/src/events/noise/tests.rs]: Dot. :)

File: exonum/src/events/noise/tests.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        // Convert to Curve25519 keys.
+        let (_, secret_key_i) = convert_keypair_for_handshake(public_key_i, secret_key_i).unwrap();
+        let (public_key_r, secret_key_r) =
+            convert_keypair_for_handshake(public_key_r, secret_key_r).unwrap();
+
+        let mut h_i = NoiseBuilder::new(PATTERN.parse().unwrap())
+            .local_private_key(&secret_key_i[..PUBLIC_KEY_LENGTH])
+            .remote_public_key(&public_key_r[..])
+            .build_initiator()
+            .unwrap();
```
Comment (popzxc) [inline exonum/src/events/noise/tests.rs]: Maybe change `unwrap`s here and below to `expect`?

File: exonum/src/events/noise/tests.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+mod test {
+    use snow::NoiseBuilder;
+    use snow::types::Dh;
+    use snow::wrappers::crypto_wrapper::Dh25519;
+
+    use crypto::PUBLIC_KEY_LENGTH;
+    use crypto::{convert_keypair_for_handshake, gen_keypair};
+
+    #[test]
+    fn test_convert_ed_to_curve_dh() {
```
Comment (popzxc) [inline exonum/src/events/noise/tests.rs]: Don't you think that this test will look better in `crypto.rs`? It's not about noise proto, it's about conversion logic.

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,11 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
 
 ## [Unreleased]
 
+### exonum
+
+- `exonum::crypto::into_x25519_keypair` method to convert from Ed25519
+  keys to X25519 keys is introduced.
```
Comment (popzxc) [inline CHANGELOG.md]: `has been introduced` to be consistent.

File: exonum/src/crypto.rs [hunk 5]
```diff
@@ -154,6 +155,36 @@ pub fn verify(sig: &Signature, data: &[u8], pubkey: &PublicKey) -> bool {
...
 }
 
+/// Converts Ed25519 keys to Curve25519.
+///
+/// Ed25519 keys used for signatures can be converted to Curve25519 and used for
+/// Diffie-Hellman key exchange.
+///
+/// # Examples
+///
+/// The example below generate a pair of secret and public Ed25519 keys and
```
Comment (popzxc) [inline exonum/src/crypto.rs]: `generates`

File: exonum/src/crypto.rs [hunk 6]
```diff
@@ -525,6 +556,28 @@ implement_private_sodium_wrapper! {
...
 
+implement_public_sodium_wrapper! {
+/// Curve25519 public key used in key exchange.
+/// This key cannot be directly generated and can only be converted
+/// from Ed25519 `PublicKey`.
+///
+/// See: [`into_x25519_keypair()`][1]
+///
+/// [1]: fn.into_x25519_keypair.html
+    struct PublicKeyX25519, PublicKeySodium, PUBLIC_KEY_LENGTH
```
[the remark this answers is not in the document]
Comment (popzxc) [reply exonum/src/crypto.rs]: Won't it end in creating a separate crate which will contain wrappers for sodiumoxide (which contains wrappers for libsodium-sys (which contains wrappers for libsodium))).

"We can solve any problem by introducing an extra level of indirection, except for the problem of too many levels of indirection".

**Comment by pavel-mukhanov** [comment]
Added 'Work in progress' again, because I need to create another patch to `exonum_sodiumoxide`.

File: CHANGELOG.md [hunk 7]
```diff
@@ -12,6 +12,13 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
 - `storage::base_index` module has become private along with `BaseIndex` and
   `BaseIndexIter` types. (#723)
 
+### New features
+
+#### exonum
+
+- `exonum::crypto::x25519::into_x25519_keypair` method to convert from Ed25519
```
Comment (popzxc) [inline CHANGELOG.md]: Maybe "`exonum::crypto::x25519` module", since now it's more than a single method.

**Pull request merged**